core/src/crds_trie.rs

```diff
@@ -0,0 +1,359 @@
+use crate::crds::VersionedCrdsValue;
+use crate::crds_gossip_pull::CrdsFilter;
+use crate::crds_value::CrdsValueLabel;
+
+/// Most significant bit in u64.
+const MSB_U64: u64 = 1u64 << 63;
+
+#[derive(Clone)]
+pub enum CrdsTrie {
+    // Left subtree, size of subtree (num of values at leaves), right subtree.
+    // Size is used in pruning to drop empty branches.
+    Node(Link, usize, Link),
+    // Since values are placed at hashes, collisions should be rare, and so a
+    // vector is more efficient than a (hash) set.
+    Leaf(Vec<CrdsValueLabel>),
+}
+
+type Link = Option<Box<CrdsTrie>>;
+
+fn new_link(leaf: bool) -> Link {
+    let node = if leaf {
+        CrdsTrie::Leaf(vec![])
+    } else {
+        CrdsTrie::Node(None, 0, None)
+    };
+    Some(Box::<CrdsTrie>::new(node))
+}
+
+/// Unboxes a &Option<Box<CrdsTrie>> to Option<&CrdsTrie>.
+#[inline]
+fn unbox_link(link: &Link) -> Option<&CrdsTrie> {
+    match link {
+        None => None,
+        Some(node) => Some(&*node),
+    }
+}
+
+impl Default for CrdsTrie {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl CrdsTrie {
+    pub fn new() -> Self {
+        Self::Node(None, 0, None)
+    }
+
+    #[must_use]
+    pub fn insert(&mut self, value: &VersionedCrdsValue) -> bool {
+        self.insert_node(
+            CrdsFilter::hash_as_u64(&value.value_hash),
+            value.value.label(),
+        )
+    }
+
+    #[must_use]
+    pub fn remove(&mut self, value: &VersionedCrdsValue) -> bool {
+        self.remove_node(
+            CrdsFilter::hash_as_u64(&value.value_hash),
+            &value.value.label(),
+        )
+    }
+
+    /// Returns all the values where the mask_bits most significant bits of
+    /// their hash is the same as the given bit-mask.
+    pub fn find(&self, mask: u64, mask_bits: u32) -> impl Iterator<Item = &CrdsValueLabel> {
+        let node = self.find_node(mask, mask_bits);
+        CrdsTrieIterator::new(node).map(|v| v.iter()).flatten()
+    }
+
+    /// Traverses the trie following the bit pattern of the given hash value
+    /// and inserts the value at the final leaf node. Returns false if the trie
+    /// does not satisfy required invariants (which should indicate a bug).
```

Comment:
Is there an advantage to returning a bool to signal correctness rather than just panicking whenever we run into an error here? It seems like we're asserting on the correctness value from the callers of this function anyway.

Reply:
one example is that the call-site can spit out debugging info before panicking (or, do clean-ups before crashing).
in general the call-site has more context than the code here, so the caller can choose to return an error, or ignore the failure and skip, or generate some diagnosis logs before panicking. so generally i think it would be better not to panic in library code.